frontend/app/src/composables/history/filter-paginate.ts

```diff
@@ -0,0 +1,201 @@
+    }
+
+    return {
+      ...selectedFilters,
+      ...get(extraParams),
+      ...get(customPageParams),
+      limit: itemsPerPage,
+      offset,
+      orderByAttributes: sortBy?.length > 0 ? sortBy : ['timestamp'],
+      ascending: sortBy?.length > 0 ? sortDesc.map(bool => !bool) : [false]
+    };
+  });
+
+  const { isLoading, state, execute } = useAsyncState<
+    Collection<V>,
+    MaybeRef<U>[]
+  >(fetchAssetData, defaultCollectionState(), {
+    immediate: false,
+    resetOnExecute: false,
+    delay: 0
+  });
+
+  const applyRouteFilter = () => {
+    if (!get(mainPage)) {
+      return;
+    }
+
+    const query = get(route).query;
+    const parsedOptions = RouterPaginationOptionsSchema.parse(query);
+    const parsedFilters = RouteFilterSchema.parse(query);
+
+    onUpdateFilters?.call(null, query);
+
+    updateFilter(parsedFilters);
+    set(paginationOptions, {
+      ...get(paginationOptions),
+      ...parsedOptions
+    });
+  };
+
+  const getQuery = (): LocationQuery => {
+    const opts = get(paginationOptions);
+    assert(opts);
+    const { itemsPerPage, page, sortBy, sortDesc } = opts;
+
+    const selectedFilters = get(filters);
+
+    const overview = get(locationOverview);
+    if (overview) {
+      selectedFilters.location = overview;
+    }
+
+    return {
+      itemsPerPage: itemsPerPage.toString(),
+      page: page.toString(),
+      sortBy,
+      sortDesc: sortDesc.map(x => x.toString()),
+      ...selectedFilters,
+      ...get(extraParams)
+    };
+  };
+
+  const fetchData = async (): Promise<void> => {
+    await execute(0, pageParams);
+  };
+
+  const setPage = (page: number) => {
+    set(userAction, true);
+    set(paginationOptions, { ...get(paginationOptions), page });
+  };
+
+  const setOptions = (newOptions: TablePagination<T>) => {
+    set(userAction, true);
+    set(paginationOptions, newOptions);
+  };
+
+  const setFilter = (newFilter: UnwrapRef<typeof filters>) => {
+    set(userAction, true);
+    updateFilter(newFilter);
+  };
+
+  onBeforeMount(() => {
+    applyRouteFilter();
+  });
+
+  watch(route, () => {
+    set(userAction, false);
+    applyRouteFilter();
+  });
+
+  watch(filters, async (filters, oldFilters) => {
+    if (isEqual(filters, oldFilters)) {
+      return;
+    }
+
+    set(paginationOptions, { ...get(paginationOptions), page: 1 });
+  });
+
+  watch(pageParams, async (params, op) => {
+    if (isEqual(params, op)) {
+      return;
+    }
+    if (get(userAction) && get(mainPage)) {
+      // Route should only be updated on user action otherwise it messes with
+      // forward navigation.
+      await router.push({
+        query: getQuery()
+      });
+      set(userAction, false);
+    }
+
+    await fetchData();
+  });
+
+  return {
+    options: paginationOptions,
+    selected,
+    openDialog,
+    editableItem,
+    itemsToDelete,
+    confirmationMessage,
+    expanded,
+    isLoading,
+    state,
+    filters,
+    matchers,
```

Comment:
Is it possible to make it so that filters is not `Ref<any>` and matches not `ComputedRef<any>`

Reply:
Sure, checking.

emphasis on `any` right?